google/cloud/pubsublite/internal/alarm_registry_impl.cc

```diff
@@ -0,0 +1,78 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsublite/internal/alarm_registry_impl.h"
+#include "google/cloud/version.h"
+#include "absl/memory/memory.h"
+#include <utility>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+AlarmRegistryImpl::AlarmRegistryImpl(google::cloud::CompletionQueue cq)
+    : cq_{std::move(cq)} {}
+
+void AlarmRegistryImpl::OnAlarm(CompletionQueue cq,
+                                std::chrono::milliseconds period,
+                                std::function<void()> const& on_alarm,
+                                std::shared_ptr<bool> const& alarm_status,
+                                std::shared_ptr<std::mutex> const& mu) {
+  {
+    std::lock_guard<std::mutex> g{*mu};
+    if (!*alarm_status) return;
+  }
+  cq.MakeRelativeTimer(period).then(
+      [cq, period, on_alarm, alarm_status,
+       mu](future<StatusOr<std::chrono::system_clock::time_point>> f) {
+        bool is_ok = f.get().ok();
```

Comment:
you don't actually need to set alarm_status to false: you're the only one looking at it. Just call:

```if (!f.get().ok()) return;```

Reply:
maybe add a LOG(INFO) as well?